onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -0,0 +1,721 @@
+      return;
+    }
+    total_channels += input_channels;
+    nchwc_inputs.push_back(it->second.get());
+  }
+
+  // Update the node to directly use the NCHWc inputs directly and decrement
+  // the original use counts of the NCHWc inputs.
+  for (size_t n = 0; n < input_defs_count; n++) {
+    input_defs[n] = nchwc_inputs[n]->nchwc_arg_;
+    nchwc_inputs[n]->remaining_original_uses_--;
+  }
+
+  // Copy the shape from any of the NCHWc inputs, but use the current node for
+  // the channel dimension.
+  NchwcArgument::Shape output_shape = nchwc_inputs[0]->shape_;
+  output_shape.dims_[1] = output_defs[0];
+
+  CreateNchwcArgument(node, node, total_channels, output_shape);
+}
+
+void NchwcTransformerImpl::TransformActivation(Node& node) {
+  auto& input_defs = node.MutableInputDefs();
+
+  auto it = nchwc_args_.find(input_defs[0]);
+  if (it != nchwc_args_.end()) {
+    auto& nchwc_input = it->second;
+    input_defs[0] = nchwc_input->nchwc_arg_;
+    nchwc_input->remaining_original_uses_--;
+
+    // Check if this is a single use NCHWc convolution that hasn't already
+    // been fused with another activation.
+    auto& nchwc_node = nchwc_input->output_node_;
+    if ((nchwc_node.OpType() == "Conv") && (nchwc_node.Domain() == kMSNchwcDomain) &&
+        (nchwc_input->starting_original_uses_ == 1) &&
+        (GetAttribute(nchwc_node, "activation") == nullptr)) {
+      nchwc_node.AddAttribute("activation", node.OpType());
+      FuseNchwcArgument(node, *nchwc_input);
+      removed_nodes_.push_front(node.Index());
+    } else {
+      CreateNchwcArgument(node, node, nchwc_input->channels_, nchwc_input->shape_);
+    }
+  }
+}
+
+void NchwcTransformerImpl::TransformElementwise(Node& node) {
+  auto& input_defs = node.MutableInputDefs();
+
+  auto it = nchwc_args_.find(input_defs[0]);
+  if (it != nchwc_args_.end()) {
+    auto& nchwc_input = it->second;
+    input_defs[0] = nchwc_input->nchwc_arg_;
+    nchwc_input->remaining_original_uses_--;
+    CreateNchwcArgument(node, node, nchwc_input->channels_, nchwc_input->shape_);
+  }
+}
+
+void NchwcTransformerImpl::Transform(Node& node) {
+  if (graph_utils::IsSupportedOptypeVersionAndDomain(node, "Conv", {1}) ||
+      graph_utils::IsSupportedOptypeVersionAndDomain(node, "FusedConv", {1}, kMSDomain)) {
+    TransformConv(node);
+  } else if (graph_utils::IsSupportedOptypeVersionAndDomain(node, "MaxPool", {1, 8, 10}) ||
+             graph_utils::IsSupportedOptypeVersionAndDomain(node, "AveragePool", {1, 7, 10}) ||
+             graph_utils::IsSupportedOptypeVersionAndDomain(node, "GlobalMaxPool", {1}) ||
+             graph_utils::IsSupportedOptypeVersionAndDomain(node, "GlobalAveragePool", {1})) {
+    TransformPool(node);
+  } else if (node.GetInputEdgesCount() == 0 && node.InputDefs().size() != 0) {
+    // The following transforms only run when the input edge count has already
+    // been decremented to zero by earlier transforms. This is a quick check
```

Comment:
What happens if the IsSupportedOptypeVersionAndDomain tests aren't satisfied? Not clear to me whether a node that has no remaining input edges needs any special handling.

Reply:
I can add more comments. The check is a quick test that the node has any reason to be checked at all with the various op name/domain/version checks. The other transforms for Add/Conv/etc only do work when all inputs are already in NCHWc format and hence have no edges present.

If the op name/domain/version checks fail, then the node will still have its input defs referring to the original NCHW NodeArg and the NchwcArgument will still have remaining_original_uses. This cases Finalize to insert ReorderOutput nodes.

Reply:
Some more comments would be great, especially connecting this with what happens in Finalize.

Reply:
Added more comments.